File: app/src/main/java/org/neidhardt/dynamicsoundboard/soundcontrol/PauseSoundOnCallListener.java
package org.neidhardt.dynamicsoundboard.soundcontrol;

import android.content.Context;
import android.telephony.PhoneStateListener;
import android.telephony.TelephonyManager;
import org.neidhardt.dynamicsoundboard.DynamicSoundboardApplication;
import org.neidhardt.dynamicsoundboard.mediaplayer.EnhancedMediaPlayer;
import org.neidhardt.dynamicsoundboard.soundmanagement.model.SoundsDataAccess;
import org.neidhardt.dynamicsoundboard.soundmanagement.service.ServiceManagerFragment;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;


public class PauseSoundOnCallListener extends PhoneStateListener
{
	private List<EnhancedMediaPlayer> pauseSounds;
	private SoundsDataAccess soundsDataAccess;

	public PauseSoundOnCallListener()
	{
		DynamicSoundboardApplication.getApplicationComponent().inject(this);
		this.pauseSounds = new ArrayList<>();
	}

	@Override
	public void onCallStateChanged(int state, String incomingNumber)
	{
		super.onCallStateChanged(state, incomingNumber);

		if (state == TelephonyManager.CALL_STATE_RINGING)
		{
			Set<EnhancedMediaPlayer> currentlyPlayingSounds = this.soundsDataAccess.getCurrentlyPlayingSounds();
			if (currentlyPlayingSounds.size() > 0)
			{
				List<EnhancedMediaPlayer> copyCurrentlyPlayingSounds = new ArrayList<>(currentlyPlayingSounds.size()); // copy to prevent concurrent modification exception
				copyCurrentlyPlayingSounds.addAll(currentlyPlayingSounds);

				for (EnhancedMediaPlayer sound : copyCurrentlyPlayingSounds)
					sound.pauseSound();
			}
			this.pauseSounds.addAll(currentlyPlayingSounds);
		}
		else if (state == TelephonyManager.CALL_STATE_IDLE)
		{
			for (EnhancedMediaPlayer player : this.pauseSounds)
				player.playSound();

			this.pauseSounds.clear();
		}
		super.onCallStateChanged(state, incomingNumber);
	}

	private void clearReferences()
	{
		this.pauseSounds.clear();
	}

	public static void registerListener(Context context, PauseSoundOnCallListener listener)
	{
		TelephonyManager manager = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
		if (manager != null)
			manager.listen(listener, PhoneStateListener.LISTEN_CALL_STATE);
	}

	public static void unregisterListener(Context context, PauseSoundOnCallListener listener)
	{
		listener.clearReferences();
		TelephonyManager manager = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
		if (manager != null)
			manager.listen(listener, PhoneStateListener.LISTEN_NONE);
	}
}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/soundcontrol/PauseSoundOnCallListener.kt
package org.neidhardt.dynamicsoundboard.soundcontrol

import android.content.Context
import android.telephony.PhoneStateListener
import android.telephony.TelephonyManager
import org.neidhardt.dynamicsoundboard.SoundboardApplication
import org.neidhardt.dynamicsoundboard.mediaplayer.MediaPlayerController
import org.neidhardt.dynamicsoundboard.notifications.NotificationService
import org.neidhardt.dynamicsoundboard.soundactivity.SoundActivity
import java.util.*


class PauseSoundOnCallListener : PhoneStateListener() {

	private val pauseSounds: MutableList<MediaPlayerController> = ArrayList()
	private val soundsDataAccess = SoundboardApplication.soundLayoutManager

	override fun onCallStateChanged(state: Int, incomingNumber: String?) {
		super.onCallStateChanged(state, incomingNumber)

		if (state == TelephonyManager.CALL_STATE_RINGING) {
			val currentlyPlayingSounds = this.soundsDataAccess.currentlyPlayingSounds
			if (currentlyPlayingSounds.isNotEmpty()) {
				val copyCurrentlyPlayingSounds = ArrayList<MediaPlayerController>(currentlyPlayingSounds.size) // copy to prevent concurrent modification exception
				copyCurrentlyPlayingSounds.addAll(currentlyPlayingSounds)

				for (sound in copyCurrentlyPlayingSounds)
					sound.pauseSound()
			}
			this.pauseSounds.addAll(currentlyPlayingSounds)
		} else if (state == TelephonyManager.CALL_STATE_IDLE) {
			for (player in this.pauseSounds)
				player.playSound()

			this.pauseSounds.clear()
		}
		super.onCallStateChanged(state, incomingNumber)
	}

	fun clearReferences() {
		this.pauseSounds.clear()
	}
}

fun SoundActivity.registerPauseSoundOnCallListener(listener: PauseSoundOnCallListener) {
    //val manager = this.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
    //manager.listen(listener, PhoneStateListener.LISTEN_CALL_STATE)
}

fun SoundActivity.unregisterPauseSoundOnCallListener(listener: PauseSoundOnCallListener) {
	////listener.clearReferences()
    val manager = this.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
	//manager.listen(listener, PhoneStateListener.LISTEN_NONE)
}

fun NotificationService.registerPauseSoundOnCallListener(listener: PauseSoundOnCallListener) {
	//val manager = this.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
	//manager.listen(listener, PhoneStateListener.LISTEN_CALL_STATE)
}

fun NotificationService.unregisterPauseSoundOnCallListener(listener: PauseSoundOnCallListener) { listener.clearReferences()
	//val manager = this.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
	//manager.listen(listener, PhoneStateListener.LISTEN_NONE)
}